feat: Add page number to Documents coming from PDFConverters and PreProcessor (#2932)
deepset-ai/haystack: docs/_src/api/api/file_converter.md

not unique, you can modify the metadata and pass e.g. `"meta"` to this field (e.g. [`"content"`, `"meta"`]).
In this case the id will be generated by using the content and the defined metadata.

<a id="parsr"></a>

# Module parsr

<a id="parsr.ParsrConverter"></a>

## ParsrConverter

```python
class ParsrConverter(BaseConverter)
```

> File converter that makes use of the open-source Parsr tool by axa-group.

Reviewer: ```suggestion
File converter that uss the open-source Parsr tool by axa-group.
```

(https://github.com/axa-group/Parsr).
This Converter extracts both text and tables.

> Supported file formats are: PDF, DOCX

Reviewer: ```suggestion
It supports PDF and DOCX files. 
```

<a id="parsr.ParsrConverter.__init__"></a>

#### ParsrConverter.\_\_init\_\_

```python
def __init__(parsr_url: str = "http://localhost:3001", extractor: Literal["pdfminer", "pdfjs"] = "pdfminer", table_detection_mode: Literal["lattice", "stream"] = "lattice", preceding_context_len: int = 3, following_context_len: int = 3, remove_page_headers: bool = False, remove_page_footers: bool = False, remove_table_of_contents: bool = False, valid_languages: Optional[List[str]] = None, id_hash_keys: Optional[List[str]] = None, add_page_number: bool = True)
```

**Arguments**:

> - `parsr_url`: URL endpoint to Parsr"s REST API.

Reviewer: ```suggestion
- `parsr_url`: URL endpoint to Parsr's REST API.
```

> - `extractor`: Backend used to extract textual structured from PDFs. ("pdfminer" or "pdfjs")

Reviewer: ```suggestion
- `extractor`: Backend used to extract text structured from PDFs ("pdfminer" or "pdfjs").
```

- `table_detection_mode`: Parsing method used to detect tables and their cells.
"lattice" detects tables and their cells by demarcated lines between cells.
"stream" detects tables and their cells by looking at whitespace between cells.

> - `preceding_context_len`: Number of lines before a table to extract as preceding context

Reviewer: ```suggestion
- `preceding_context_len`: The number of lines that precede a table to extract as preceding context
```

> (will be returned as part of meta data).

Reviewer: ```suggestion
(this number is returned as part of meta data).
```

> - `following_context_len`: Number of lines after a table to extract as preceding context

Reviewer: ```suggestion
- `following_context_len`: The number of lines following a table to extract as preceding context
```

> (will be returned as part of meta data).

Reviewer: ```suggestion
(this number is returned as part of meta data).
```

> - `remove_page_headers`: Whether to remove text that Parsr detected as a page header.

Reviewer: ```suggestion
- `remove_page_headers`: Removes the text that Parsr detected as a page header. Possible values: True, False.
```

> - `remove_page_footers`: Whether to remove text that Parsr detected as a page footer.

Reviewer: ```suggestion
- `remove_page_footers`: Removes the text that Parsr detected as a page footer. Possible values: True, False.
```

> - `remove_table_of_contents`: Whether to remove text that Parsr detected as a table of contents.

Reviewer: ```suggestion
- `remove_table_of_contents`: Removes the text that Parsr detected as a table of contents. Possible values: True, False.
```

> - `valid_languages`: Validate languages from a list of languages specified in the ISO 639-1

Reviewer: ```suggestion
- `valid_languages`: Validate languages from a list of languages specified in the [ISO 639-1]
```

(https://en.wikipedia.org/wiki/ISO_639-1) format.

> This option can be used to add test for encoding errors. If the extracted text is

Reviewer: ```suggestion
Use this option can add test for encoding errors. If the extracted text is
```

> not one of the valid languages, then it might likely be encoding error resulting

Reviewer: ```suggestion
not one of the valid languages, it might likely be encoding error resulting
```

in garbled text.

> - `id_hash_keys`: Generate the document id from a custom list of strings that refer to the document's

Reviewer: ```suggestion
- `id_hash_keys`: Generate the document ID from a custom list of strings that refer to the document's
```

> attributes. If you want to ensure you don't have duplicate documents in your DocumentStore but texts are

Reviewer: ```suggestion
attributes. To ensure you don't have duplicate documents in your DocumentStore but texts are
```

> not unique, you can modify the metadata and pass e.g. `"meta"` to this field (e.g. [`"content"`, `"meta"`]).

Reviewer: ```suggestion
not unique, you can modify the metadata and pass for example, `"meta"` to this field ([`"content"`, `"meta"`]).
```

> In this case the id will be generated by using the content and the defined metadata.

Reviewer: ```suggestion
If you do this, the ID is generated by using the content and the defined metadata.
```

- `add_page_number`: Whether to add the number of the page a table occurs in to the Document's meta field
`"page"`.

<a id="parsr.ParsrConverter.convert"></a>

#### ParsrConverter.convert

```python
def convert(file_path: Path, meta: Optional[Dict[str, Any]] = None, remove_numeric_tables: Optional[bool] = None, valid_languages: Optional[List[str]] = None, encoding: Optional[str] = "utf-8", id_hash_keys: Optional[List[str]] = None) -> List[Document]
```

Extract text and tables from a PDF or DOCX using the open-source Parsr tool.

**Arguments**:

- `file_path`: Path to the file you want to convert.

> - `meta`: Optional dictionary with metadata that shall be attached to all resulting documents.

Reviewer: ```suggestion
- `meta`: Optional dictionary with metadata that you want to attach to all resulting documents.
```

Can be any custom keys and values.
- `remove_numeric_tables`: Not applicable.

> - `valid_languages`: Validate languages from a list of languages specified in the ISO 639-1

Reviewer: ```suggestion
- `valid_languages`: Validate languages from a list of languages specified in the [ISO 639-1]
```

(https://en.wikipedia.org/wiki/ISO_639-1) format.

> This option can be used to add test for encoding errors. If the extracted text is

Reviewer: ```suggestion
Use this option to add test for encoding errors. If the extracted text is
```

> not one of the valid languages, then it might likely be encoding error resulting

Reviewer: ```suggestion
not one of the valid languages, it might likely be encoding error resulting
```

in garbled text.
- `encoding`: Not applicable.

> - `id_hash_keys`: Generate the document id from a custom list of strings that refer to the document's

Reviewer: ```suggestion
- `id_hash_keys`: Generate the document ID from a custom list of strings that refer to the document's
```

> attributes. If you want to ensure you don't have duplicate documents in your DocumentStore but texts are

Reviewer: ```suggestion
attributes. To ensure you don't have duplicate documents in your DocumentStore but texts are
```

> not unique, you can modify the metadata and pass e.g. `"meta"` to this field (e.g. [`"content"`, `"meta"`]).

Reviewer: ```suggestion
not unique, modify the metadata and pass, for example,  `"meta"` to this field ([`"content"`, `"meta"`]).
```

> In this case the id will be generated by using the content and the defined metadata.

Reviewer: ```suggestion
If you do this, the ID is generated by using the content and the defined metadata.
```
